### ISSUE TITLE
Weird FindFirstFileExW behavior on FAT32

### ISSUE THREAD

**Issue by iSazonov** [dotnet/runtime#64789]
In PowerShell repository we catch weird issue https://github.com/PowerShell/PowerShell/issues/16804 

In short, FindFirstFileExW can return native error 1113 "No mapping for the Unicode character exists in the target multi-byte code page." on FAT32 volume if folder/file name is Unicode like Japanese.

@SteveL-MSFT Found the discussion https://groups.google.com/g/fastcopy-bb-eng/c/z6DJRzeGeL4/m/IEsaD7g6AgAJ?pli=1
I replaced FindFirstFileExW with FindFirstFileW in PowerShell and this works as expected now.

It is very amazing FindFirstFileExW behavior - look like Windows API bug.

I report this here because I see .Net uses internally FindFirstFileExW some times. I did not investigate but this could raise issues with .Net too.
.Net team could replace FindFirstFileExW with FindFirstFileW at least or communicate with Windows team to resolve this in depth.

**Comment by danmoseley** [comment]
@JeremyKuhne  thoughts?

**Comment by JeremyKuhne** [comment]
@danmoseley `FindFirstFileW` just calls `FindFirstFileExW` with `FindExInfoStandard` and `FindExSearchNameMatch` so I doubt that this is a Windows bug.

**Comment by JeremyKuhne** [comment]
@danmoseley One should probably check to see if they are inadvertently calling the ANSI (`A`) version. That would be problematic if things couldn't be projected to the current code page.

**Comment by iSazonov** [comment]
@JeremyKuhne In OP there is a link to detailed discussion https://groups.google.com/g/fastcopy-bb-eng/c/z6DJRzeGeL4/m/IEsaD7g6AgAJ?pli=1
Also I can confirm that in PowerShell code we call exactly `FindFirstFileExW`
https://github.com/PowerShell/PowerShell/blob/bcc1f10b6712e1b898c16e056b4ca3b95f14d634/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L8076-L8077

**Comment by danmoseley** [comment]
Yes, here is the entire implementation of FindFirstFileExW, and I verified it was the same back in Windows 7:
```c
    return FindFirstFileExW(
                lpFileName,
                FindExInfoStandard,
                lpFindFileData,
                FindExSearchNameMatch,
                NULL,
                0
                );
```
This is how we call FindFirstFileExW, except we use FindExInfoBasic instead of FindExInfoStandard. I see the Powershell change to FindFirstFileW implicitly changed from FindExInfoBasic  as well https://github.com/PowerShell/PowerShell/pull/16840

Based on the symptoms above, and assuming that Powershell change actually fixed it (?) then perhaps that is relevant. But FindExInfoStandard does _more_ work.

With respect to "FastCopy" the source code at https://github.com/shirouzu/FastCopy stops before the updates mentioned in that thread. I can't really look at it anyway because it's GPL.

@iSazonov is it possible for you to supply a repro that does not involve Powershell?

**Comment by iSazonov** [comment]
> @iSazonov is it possible for you to supply a repro that does not involve Powershell?

Attached
[findfirst.zip](https://github.com/dotnet/runtime/files/8008642/findfirst.zip)

**Comment by danmoseley** [comment]
Thanks for the repro. I added a bit to compare both API: https://gist.github.com/danmoseley/36ff1adc1f7c882d57b4e655b5d55fca

If I change the FindFirstFileExW call to pass FindExInfoStandard instead of FindExInfoBasic, so it behaves like FindFirstFileW, unsurprisingly the result matches FindFirstFileW.

```
C:\proj\ff>dotnet run
Hello, World!
FindFirstFileEx: No mapping for the Unicode character exists in the target multi-byte code page.
FindFirstFile: success

C:\proj\ff>dotnet run
Hello, World!
FindFirstFileEx: success
FindFirstFile: success
```

It seems a Windows bug.

**Comment by danmoseley** [comment]
I can't figure this out without a kernel debugger. I opened a [bug against Windows](https://microsoft.visualstudio.com/OS/_workitems/edit/38065608). The link from the Powershell issue suggests it reproes back on Windows 7 too.

Meantime if we want to fix this, we either need to switch to passing FindExInfoStandard, or try again with FindExInfoStandard when we get STATUS_UNMAPPABLE_CHARACTER. Since we're using FindExInfoBasic for performance, and this is the first report of the bug, falling back to a 2nd call with FindExInfoStandard is probably best.

**Comment by danmoseley** [comment]
Internally FindFirstFile[Ex] ends up calling 
```c
    Status = NtQueryDirectoryFileEx(
        hFindFile,
        NULL,
        NULL,
        NULL,
        &IoStatusBlock,
        FullDirectoryInfo,
        sizeof(Buffer),
        (fInfoLevelId == FindExInfoStandard) ? FileBothDirectoryInformation :
        FileFullDirectoryInformation,
        FILE_QUERY_RETURN_SINGLE_ENTRY |
        ((dwAdditionalFlags & FIND_FIRST_EX_ON_DISK_ENTRIES_ONLY)
            ? FILE_QUERY_RETURN_ON_DISK_ENTRIES_ONLY
            : 0),
        &FileName
    );
```
So FindExInfoBasic leads to passing the FileFullDirectoryInformation flag.

We don't call NtQueryDirectoryFileEx directly but we do call NtQueryDirectoryFile directly, and [we pass FileFullDirectoryInformation there](https://github.com/dotnet/runtime/blob/9df283248256dd880eb1ee5ecce533ad206a55d3/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerator.Windows.cs#L94) . However this doesn't seem to be a problem as `Directory.EnumerateFiles` does see the problematic file.

**Comment by danmoseley** [comment]
It's not a problem with Directory.EnumerateFiles because of something about how the handle is created that we're passing to NtQueryDirectoryFile. If I create a handle with FindFirstFile on the parent directory, NtQueryDirectoryFileEx with FileFullDirectoryInformation does fail. But if I create a handle with CreateFile on the parent directory ([as we do)](https://github.com/dotnet/runtime/blob/9df283248256dd880eb1ee5ecce533ad206a55d3/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEnumerator.Windows.cs#L167) then it does not fail.

Our uses of FindFirstFileXXX directly are relatively obscure, so it's not clear that we're affected by the bug. It would probably be good to modify Interop.FindFirstFileEx.cs directly do add the workaround, but any test would likely have to be against Interop.Kernel32.FindFirstFile directly.

**Comment by danmoseley** [comment]
@JeremyKuhne in case you have any further thoughts about the above info.

**Comment by iSazonov** [comment]
Since there are ways without the fail this nervous me because it is about Unicode and maybe related with security, not in the scenario but in scenarios where _no_ the fail is. Perhaps it makes sense to ping security team too.

**Comment by JeremyKuhne** [comment]
@danmoseley I'd step through to see precisely what the difference is between the calls to NtOpenFile & NtQueryDirectoryFileEx to try to pin down why we succeed. I suspect this _might_ have something to do with the Unicode filename being the filter. _FindFirstFile_ splits off the filename as the filter string. It would be worth checking enumerating a directory with the same Unicode name and a wildcard `*` filter to see if there are issues with the directory name being unmappable to the current CodePage. I moved the filtering to our code with the new enumeration so we're never letting the OS filter.

As you've mentioned, there is extra work to get the 8.3 name, so I wouldn't default to that. Fallback probably makes sense as there is some optimization with UNC based requests (filtering is done on the server side).

FileSystemInfo uses FindFirstFile to get attributes when access is denied, that should be tested. The other usage in RemoveDirectoryRecursive adds a filter so it _might_ not be impacted. If it is it might be worth looking at converting it to use our new enumeration APIs.

**Comment by danmoseley** [comment]
Thanks @JeremyKuhne . Looks like this one can't be up for grabs as it would need stepping through the Windows code. I do think this is low priority unless other evidence emerges so I don't plan to look further myself.

**Comment by iSazonov** [comment]
I personally don't like to hide the bug with the fallback. If an application throws user can rename a file/folder and continue to use the application. But it is very difficult for the user to guess the reason why an application suddenly starts to slow down. 
I hope the bug will be fixed in Windows.

**Comment by danmoseley** [comment]
I do not know whether it will be fixed but it would also have to be back ported way back. If we get reports of customers actually hitting this in .NET Core then a workaround may be needed.

**Comment by iSazonov** [comment]
> I do think this is low priority unless other evidence emerges

This is unlikely to get other reports for a simple reason - FAT32 is not rarely used, but is used in specific scenarios with _end users_ so there is simply no one to report such problems. :-)

**Comment by iSazonov** [comment]
We get new message that cmd.exe doesn't work well too https://github.com/PowerShell/PowerShell/issues/16804#issuecomment-1038384989

**Comment by danmoseley** [comment]
@iSazonov presumably `Remove-Item` is implemented by calls to our API's, not by pinvokes on Powershell side? If so that means there must be some repo you can share that uses our API's and not pinvokes? We would want to fix that.
